### ISSUE TITLE
./bootstrap-vcpkg.sh fails to link vcpkg and create executable binary with latest update

### ISSUE THREAD

**Issue by timautry** [microsoft/vcpkg#10847]
**Describe the bug**
./bootstrap-vcpkg.sh fails to link vcpkg and create executable binary with latest update

**Environment**
- OS: [Linux OpenSuse Leap 15.1 - latest updates]
- Compiler: GNU 8.2.1 - latest updates

**To Reproduce**
1. git clone --verbose https://github.com/Microsoft/vcpkg.git ./vcpkg
2. cd ./vcpkg
3. ./bootstrap-vcpkg.sh
4. **[72/72] Linking CXX executable vcpkg**
**FAILED: vcpkg**
 ... see attached _**bootstrap-vcpkg.log**_ for complete error listing produced when running ./bootstrap-vcpkg.sh

**Expected behavior**
bootstrap-vcpkg.sh should be compile, link and create vcpkg

**Failure logs** 
[bootstrap-vcpkg.log](https://github.com/microsoft/vcpkg/files/4483143/bootstrap-vcpkg.log)

**Additional context**
- Initially since vcpkg sources were updated maybe the changes were incompatible with the an existing vcpkg instance, so I then created a new instance and it also produces the same failed results.
- Also tried run ./bootstrap-vcpkg.sh logged in as _root_ with same results.
- I did a _**git reset --hard 1e19af09e53e5f306ed89c2033817a21e5ee0bcf**_ on the linux instance to back out the latest two changes and ./bootstrap-vcpkg.sh compiled, linked and created the vcpkg binary properly.
- On Windows 10 x64 the latest vcpkg with all updates  .\bootstrap-vcpkg.bat compiles, links and creates vcpkg.exe as expected.

**Comment by NancyLi1013** [comment]
Hi @timautry 
Thanks for posting this issue.
It seems to be introduced by the latest commits.

@strega-nil  
I also encountered the similar issue on gcc 9.2.1.
Could you please help look into this?

**Comment by strega-nil** [comment]
I don't know how this issue is coming up; it looks like it isn't correctly linking the standard library. However, on my machine, it works. @NancyLi1013 or @timautry, can you try changing the line in `scripts/bootstrap.sh`:

```sh
("$cmakeExe" -B "$buildDir" -S toolsrc -DCMAKE_BUILD_TYPE=Release -G "Ninja" "-DCMAKE_MAKE_PROGRAM=$ninjaExe" "-DBUILD_TESTING=OFF" "-DVCPKG_DEVELOPMENT_WARNINGS=OFF" "-DVCPKG_WARNINGS_AS_ERRORS=OFF" "-DVCPKG_DISABLE_METRICS=$vcpkgDisableMetrics" "-DVCPKG_ALLOW_APPLE_CLANG=$vcpkgAllowAppleClang") || exit 1
```

to

```sh
(CXX=$CXX "$cmakeExe" -B "$buildDir" -S toolsrc -DCMAKE_BUILD_TYPE=Release -G "Ninja" "-DCMAKE_MAKE_PROGRAM=$ninjaExe" "-DBUILD_TESTING=OFF" "-DVCPKG_DEVELOPMENT_WARNINGS=OFF" "-DVCPKG_WARNINGS_AS_ERRORS=OFF" "-DVCPKG_DISABLE_METRICS=$vcpkgDisableMetrics" "-DVCPKG_ALLOW_APPLE_CLANG=$vcpkgAllowAppleClang") || exit 1
```

and rerun?

**Comment by Farwaykorse** [comment]
Like @NancyLi1013 I see a linking failure with GCC 9.2.1 - _on the WSL (Ubuntu  18.04.4 LTS)_
But this concerns a `undefined reference to std::experimental::filesystem::v1::***`
[console output](https://gist.github.com/Farwaykorse/30c05f1125314f0e24ef5d2234efd9b5)

Like @timautry vcpkg builds fine after running `git reset --hard 1e19af0` 
@strega-nil your suggested addition of `CXX=$CXX` to `scripts/bootstrap.sh` has no effect on this case. It might for @timautry

**Comment by timautry** [comment]
I cloned a new instance on the Linux machine and made the suggested change by @strega-nil to _vcpkg-root_/scripts/bootstrap.sh line 244 and when I ran the _vcpkg-root_/bootstrap-vcpkg.sh I get the following:

"tim.autry@galactica:[~/git/vcpkg.test]"
">$ ./bootstrap-vcpkg.sh"
Downloading cmake...
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   639  100   639    0     0   1166      0 --:--:-- --:--:-- --:--:--  1166
100 35.4M  100 35.4M    0     0  20.4M      0  0:00:01  0:00:01 --:--:-- 45.1M
Downloading cmake... done.
Extracting cmake...
Extracting cmake... done.
Downloading ninja...
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   624  100   624    0     0   2000      0 --:--:-- --:--:-- --:--:--  2000
100 77854  100 77854    0     0   122k      0 --:--:-- --:--:-- --:--:--  873k
Downloading ninja... done.
Extracting ninja...
Extracting ninja... done.
**Error: could not load cache**


That is the only message I get.

Note: 1. The linux instance of vcpkg is on Hyper-V machine on Windows 10 Pro x64, loaded from the latest OpenSuse-Leap 15.1 DVD with all current updates to the OS and GNU C/C++ 8.2.1.

**Comment by aspioupiou** [comment]
We noticed the exact same behaviour on Ubuntu 18.04.1 with gcc9.2.1

**Comment by strega-nil** [comment]
Alright, could someone try running the cmake manually?

```sh
$ cd toolsrc
$ mkdir out
$ cmake ..
$ cmake --build .
```

**Comment by strega-nil** [comment]
@Farwaykorse okay that should actually never happen, could you please run the following:

```sh
cat >test.cxx <<EOF
#include <iostream>
int main() {
  std::cout << _GLIBCXX_RELEASE << '\n';
}
EOF
c++ test.cxx
./a.out
```

**Comment by strega-nil** [comment]
BTW, I appreciate y'all helping me so much, thanks!

**Comment by Farwaykorse** [comment]
> @Farwaykorse okay that should actually never happen, could you please run the following:
> ...

```Bash
farwaykorse@P53:~/tmp$ cat >test.cxx <<EOF
> #include <iostream>
> int main() {
>   std::cout << _GLIBCXX_RELEASE << '\n';
> }
> EOF
farwaykorse@P53:~/tmp$ c++ test.cxx
farwaykorse@P53:~/tmp$ ./a.out
9
```

**Comment by timautry** [comment]
@strega-nil See attached newtest.txt. I had a few issues:
- I had to use "cmake ." instead of "cmake .."
- then I had to add the path of the cmake downloaded/installed in _vcpkg-root_/downloads/tools/cmake-3.14.0-linux/cmake-3.14.0-Linux-x86_64/bin by bootstrap-vcpkg.sh because the default cmake installed by OpenSuse Leap 15.1 is only 3.10.2 which is less than what vcpkg requires. 
- It still did not link
[newtest.txt](https://github.com/microsoft/vcpkg/files/4488667/newtest.txt)

**Comment by strega-nil** [comment]
@timautry sorry, could you `cd` into `out` first, missed that.

**Comment by strega-nil** [comment]
@Farwaykorse could you run the following then:

```sh
cat toolsrc/build.rel/CMakeFiles/CMakeError.log
```

**Comment by Farwaykorse** [comment]
For me it builds with:
````
$ cd toolsrc
$ mkdir out
$ cd out
$ cmake ..
$ cmake --build .
````
(same when using `-G Ninja`)

**Comment by Farwaykorse** [comment]
> @Farwaykorse could you run the following then:
> 
> ```shell
> cat toolsrc/build.rel/CMakeFiles/CMakeError.log
> ```

After rerunning `.\bootstrap-vcpkg.sh`.
[toolsrc/build.rel/CMakeFiles/CMakeError.log](https://gist.github.com/Farwaykorse/d271eb12f348c47cc61cafa0020a1d37)

**Comment by timautry** [comment]
@strega-nil cc: @Farwaykorse 
Ok it did not give me any errors back out to console, but it just stopped at linking 79% - see attached.
[newtest1.log](https://github.com/microsoft/vcpkg/files/4488837/newtest1.log)
[CMakeError.log](https://github.com/microsoft/vcpkg/files/4488839/CMakeError.log)
[CMakeOutput.log](https://github.com/microsoft/vcpkg/files/4488840/CMakeOutput.log)

**Comment by strega-nil** [comment]
Alright then; I'm going to revert the changes. Is there any way you two would be willing to continue trying this with changesets, because I would like to un-revert the changes at some point.

**Comment by Farwaykorse** [comment]
That's fine.

**Comment by BillyONeal** [comment]
I observe that the default compiler from `zypper install gcc` is 7.0 but cmake here is choosing 8.0, could that be related?

**Comment by BillyONeal** [comment]
Is it possible that we built some stuff with gcc-8 but `/usr/bin/g++` is gcc-7?

**Comment by timautry** [comment]
Because the bootstrap-vcpkg.sh uses the symlinks called g++ and gcc in the /usr/bin.
I use a process call alternatives to quickly change between versions of g++ and gcc
- see attached 
[gcc update-alternatives - openSUSE Wiki.pdf](https://github.com/microsoft/vcpkg/files/4489429/gcc.update-alternatives.-.openSUSE.Wiki.pdf)

**Comment by linquize** [comment]
Why check compiler version?
Should perform feature detection instead of detecting version number.
Should try to compile with and without libstdc++fs
Should try to compile with and without <experimental/filesystem>

**Comment by NancyLi1013** [comment]
@strega-nil 
Sorry for the late response.
I have tried on Ubuntu 16.04 gcc 8.3.0 and Ubuntu 18.04 gcc 9.2.1 based on the PR #10867 and there are no any errors.
It seems to be normal work now.

**Comment by past-due** [comment]
@strega-nil: I'm seeing issues building vcpkg with `g++-9` on macOS 10.14. It seems the added

```sh
if which g++ >/dev/null 2>&1; then
   CXX=g++
```

lines at the top of `selectCXX()` function (ref: https://github.com/microsoft/vcpkg/commit/22623e35016cae6061e0d9e502577077b3c33fd9)

are causing the detection of `/Applications/Xcode_10.3.app/Contents/Developer/usr/bin/g++`, instead of the Homebrew-installed `g++-9`.

**Comment by strega-nil** [comment]
@past-due yeah; try out #10867 to fix this. I'm going to merge that once the CI finishes running.

**Comment by ASxa86** [comment]
Has the reverted change made its way to master yet? I am currently seeing this problem on Ubuntu 19.10 using gcc/g++ v9.2.1. Is there a quick fix I could do in the mean time that I'm missing?

**Comment by strega-nil** [comment]
Fixed by #10867.

**Comment by past-due** [comment]
@strega-nil: Just confirming that this fixes the issues on macOS as well. 👍
